lib/tlog/tty_source.c

```diff
@@ -274,10 +277,18 @@ tlog_tty_source_read(struct tlog_source *source, struct tlog_pkt *pkt)
                 if (clock_gettime(tty_source->clock_id, &ts) < 0) {
                     return TLOG_GRC_ERRNO;
                 }
-                tlog_pkt_init_io(pkt, &ts,
-                                 tty_source->fd_idx ==
-                                    TLOG_TTY_SOURCE_FD_IDX_OUT,
-                                 tty_source->io_buf, false, rc);
+                tlog_pkt_init_io(pkt, &ts, output, tty_source->io_buf, false, rc);
+            } else if (rc == 0) {
+                tty_source->fd_list[tty_source->fd_idx].fd = -1;
+
+                /* Input:  Initialize and return a zero-length packet indicating EOF.
+                 *         No remaining I/O from this file descriptor
+                 * Output: Return a Void packet, end of source
+                 *         data
+                 */
+                if (tty_source->fd_list[TLOG_TTY_SOURCE_FD_IDX_OUT].fd >= 0) {
```

Comment:
I'm still struggling with understanding this. I'll try to talk through it:

So, if the read returns zero data, it means we've got to EOF on that FD. So, we mark that FD as such by assigning an invalid FD value to its index (could be useful to explain that step as well). Then we check that the output FD has not got the EOF yet. That means that we just got an EOF on the input FD. So, first question: why don't we just check that we're handling the input FD? I.e. why don't we write it more directly:

```c
if (!output) {
    tlog_pkt_init_eof(pkt, output);
}
```

Next, I don't understand **why** we have to return the EOF packet for input, but the void packet for output? Could you explain *that*?

Reply:
HI @spbnick no problem, I prefer to wait until we get these concerns addressed before pushing this.

First question: Yes, that is more direct to use the code you pasted. I will update the PR to use that approach.

Second question: We return an EOF packet for input because we still need to transfer and consume output from the recorded process after input read returns zero data. If we return a Void packet for input then we will prematurely break out of the main transfer loop before reading output I/O.

If we wait for input **and** output to close before returning a Void packet, then tlog will hang in instances where no input data is available because a poll event is never generated for input, and thus we never reach the code to 'close' the input I/O File descriptors and invalidate the input FD index.

Therefore the logic that has worked in my testing is to continue transferring I/O after a read returns 0 for input, but return a Void packet to finish and break out of transferring I/O after output read returns 0. This may be problematic if both input and output I/O is available and the  TTY source poll() triggers generates an output FD event **before** an input event because then we will miss the input I/O entirely. This did not occur in my testing, or with the ansible pipelining case so I'm not sure how to best handle this case.

Hope this helps.

Reply:
Thank you for the explanations, @justin-stephenson. However, I think we strayed off path here quite a bit.

Take a look at the two programs below.

`shut_and_wait`:
```bash
#!/bin/bash

echo "Shutting up!"
# Close all I/O
exec 0<&- 1>&- 2>&-
# Shouldn't block
cat
# Shouldn't print anything
echo "Blah"
echo "Blah" >&2
sleep 5
```

`ssh_shut_up`:
```bash
#!/bin/bash

# Close all I/O
exec 0<&- 1>&- 2>&-
ssh localhost ./shut_and_wait
```
If you put the first one (`shut_and_wait`) into your home folder, and then execute `ssh localhost ./shut_and_wait`, you will note that it only prints the `Shutting up!` message, and doesn't print anything else, nor does it block on `cat`, because all inputs and outputs are closed, and yet `ssh` doesn't finish until the program exits after 5 seconds. This proves that `ssh` doesn't exit until the child exits and no I/O conditions matter. Moreover, if you execute `ssh_shut_up` you will see that it also waits 5 seconds, even though all the I/O file descriptors on the **host** are closed too.

So, it's good that we're fixing the `tty_source` to not return the void packet on any files being closed and not making tlog exit, but we should be going further and not exiting until the child exits, no matter which file descriptors are closed. However, we should still track which FDs are closing and close the corresponding end to communicate properly.

So, I think `tty_source` should simply be returing an EOF packet for **either** input and output EOF, and only ever return the void packet if it cannot return any data at all, i.e. when **all** FDs are closed and `win_fd` is invalid (that is we're not reporting window resizes). At the same time, `tlog_rec_transfer` shouldn't be exiting even if it gets the void packet from the `tty_source`, but only if it detects that the child has exited. I'll leave it to you to research how to best do the latter.

Reply:
PR Updated based on your comments here, hopefully this version looks better. 

One comment that might be worth mentioning:
- Exiting the `tlog_rec_transfer` loop immediately when **SIGCHLD** is caught results in incomplete I/O transferred(including with SA_RESTART), therefore I use the `read_pending` conditional, separate **SIGCHLD** handler, and the below code to complete transferring all I/O before exiting.
~~~
+        /* Exit the loop */
+        if (tlog_rec_child_exited && !read_pending) {
+            break;
+        }
~~~

Thank you again for the insight and review Nikolai.

Reply:
Right, I think we should go even further than that. Please see my review comments.